### PR TITLE
Use dch instead of hack to generate changelog

### PR DESCRIPTION
**Pull request by npostavs**
**Short description of changes**

Use dch to generate Debian changelog instead of handrolled shell calls

**Context: Fixes an issue?**

This is an alternate version of #1767

**Does this change need documentation? What needs to be documented and how?**

No.

**Status of this Pull Request**

Working implementation, a sample output is [changelog.dch.log](https://github.com/jamulussoftware/jamulus/files/7620907/changelog.dch.log). (updated with Nov 29th changes)

**What is missing until this pull request can be merged?**

~~Possibly the "initial release" line needs to be removed somehow.~~

## Checklist

- [X] I've verified that this Pull Request follows the [general code principles](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#jamulus-projectsource-code-general-principles)
- [X] I tested my code and it does what I want
- [X] My code follows the [style guide](https://github.com/jamulussoftware/jamulus/blob/master/CONTRIBUTING.md#source-code-consistency) 
- [X] I waited some time after this Pull Request was opened and all GitHub checks completed without errors. 
- [X] I've filled all the content above

### CODE REVIEW

**Comment by softins** [comment]
Firstly, thank you @npostavs for addressing the build warnings about the `debian/changelog` entry errors, which had been bothering me for some time. However, I think the proposed change is incomplete, as it introduces other issues:

1. The modification to `.github/actions_scripts/getChangelog.pl` changes the output from MarkDown to a single line per change.

    This is certainly useful for passing each line to `dch`, but the `getChangeLog.pl` script is also used in `.github/actions_scripts/analyse_git_reference.py` to create a MarkDown file containing the version release notes. `analyse_git_reference.py` is called from `.github/workflows/autobuild.yml` in the section `create_release`.

    The new output format will not be compatible with this requirement.

    Maybe instead leave `getChangeLog.pl` unchanged, and create instead a new script (e.g. `getChangeLines.pl`?) for use with `dch`, which could be called from `distributions/build-debian-package-auto.sh`? Or else provide a parameter to the script?

2. The `debian/changelog` produced by the new version of `build-debian-package-auto.sh` changes the maintainer information from what was used previously. After some experimentation, I found that I preferred the output if I changed the `dch` lines as follows:

    ```
    DEBFULLNAME=GitHubActions DEBEMAIL=noemail@example.com dch --nomultimaint --newversion "${VERSION}" ''
    ```

    (or some appropriate values - the above matches the old script) and:

    ```
    dch --nomultimaint "$entry"
    ```

**Comment by softins** [comment]
After further experimentation, this is the current state of my `build-debian-package-auto.sh`:

```
#!/bin/sh -e

# Create deb files

cp -r debian ..
cd ..

# get the jamulus version from pro file
VERSION=$(cat Jamulus.pro | grep -oP 'VERSION = \K\w[^\s\\]*')

export DEBFULLNAME=GitHubActions DEBEMAIL=noemail@example.com

# Generate Changelog
echo -n generating changelog
rm -f debian/changelog
dch --create --package jamulus --empty --newversion "${VERSION}" ''
perl .github/actions_scripts/getChangelog.pl ChangeLog "${VERSION}" | while read entry
do
    echo -n .
    dch "$entry"
done
echo

echo "${VERSION} building..."

sed -i "s/é&%JAMVERSION%&è/${VERSION}/g" debian/control

debuild --preserve-env -b -us -uc
```

It uses `--create`, `--package`, `--newversion` and `--empty` to create a new `debian/changelog` file. The use of `--empty` causes a warning, which the man page says is expected, but it does avoid the spurious "Initial version" log entry.

It exports `DEBFULLNAME` and `DEBEMAIL` so they are available to all calls of `dch`. This also means `--nomultimaint` is not necessary.

**Comment by npostavs** [comment]
> 1. but the `getChangeLog.pl` script is also used in `.github/actions_scripts/analyse_git_reference.py`

Ah, I missed that.

> Or else provide a parameter to the script?

I went with this option.

> After further experimentation, this is the current state of my `build-debian-package-auto.sh`:

Added to the PR.

**Comment by ann0see** [comment]
@npostavs what is the Next action on this? Review, testing?

**Comment by npostavs** [comment]
Yeah, I believe the code is ready, and I have addressed all of @softin's concerns.

File: .github/actions_scripts/getChangelog.pl [hunk 1]
```diff
@@ -2,18 +2,34 @@
...
 open my $fh, '<', $ChangeLogFile or die;
-$/ = undef;
-my $ChangeLog = <$fh>;
-close $fh;
 
-if ( $ChangeLog =~ m/^### $Version (.*?)$(.*?)^###/sm ) {
-    print $2;
-} else {
-    print "No changelog found for this version.";
+my $SingleLineEntries = ($FormatOption or '') ne '--line-per-entry';
```
Comment (softins) [inline .github/actions_scripts/getChangelog.pl:10]: I've just tested this and it appears to be the wrong way up.
```suggestion
my $SingleLineEntries = ($FormatOption or '') eq '--line-per-entry';
```

**Review by softins — CHANGES_REQUESTED** [review]
Only the one change required. Everything else looks ok, and I have successfully build a deb.

File: .github/actions_scripts/getChangelog.pl (hunk 1 above)
Comment (npostavs) [reply .github/actions_scripts/getChangelog.pl]: Oh, I had fixed that in my working copy where I was testing, but forgot to commit it. Done now.

File: distributions/build-debian-package-auto.sh [hunk 2]
```diff
@@ -2,30 +2,24 @@
...
-echo "" >> debian/changelog
-echo " -- GitHubActions <noemail@example.com> ${DATE} +0001" >> debian/changelog
-echo "" >> debian/changelog
-cat distributions/debian/changelog >> debian/changelog
-
-# patch the control file
-# cp the modified control file here
-
-cp distributions/autobuilddeb/control debian/control
+export DEBFULLNAME=GitHubActions DEBEMAIL=noemail@example.com
```
Comment (ann0see) [inline distributions/build-debian-package-auto.sh:11]: Concerning the E-Mail, is there a standard mail saying basically "we don’t publish an e-mail address"?

Comment (ann0see) [inline distributions/build-debian-package-auto.sh:11]:
```suggestion
export DEBFULLNAME=JamulusAutobuild DEBEMAIL=noreply@example.com
```

**Review by ann0see — COMMENTED** [review]
Just a minor thought. It’s still not optimal, but probably better then now.

File: distributions/build-debian-package-auto.sh (hunk 2 above)
Comment (softins) [reply distributions/build-debian-package-auto.sh:11]: The original corresponds to what was in the old method. I would be inclined not to change that in this PR, but to consider it separately.

**Review by softins — APPROVED** [review]
Looks fine now, thank you!

**Review by ann0see — COMMENTED** [review]
Did you also test if the automated change log detection for GH releases still works? Seems ok, but better test it.

Urgh. Didn't read... Should be ok.

File: .github/actions_scripts/getChangelog.pl [hunk 3]
```diff
@@ -2,18 +2,34 @@
...
 
-if ( $ChangeLog =~ m/^### $Version (.*?)$(.*?)^###/sm ) {
-    print $2;
-} else {
-    print "No changelog found for this version.";
+my $SingleLineEntries = ($FormatOption or '') eq '--line-per-entry';
+
+my $logversion = "";
+my $entry;
+while (my $line = <$fh>) {
```
Comment (ann0see) [inline .github/actions_scripts/getChangelog.pl:14]: I *think* we should stop looping through the file if we already found the correct version ($logversion changed to the next version). Then we could also check if a version wasn't found as soon as we reached the EOF and output some error message.

Comment (ann0see) [reply .github/actions_scripts/getChangelog.pl:14]: @npostavs could you please insert a break statement to stop looping through unneeded lines?

**Review by ann0see — CHANGES_REQUESTED** [review]

File: .github/actions_scripts/getChangelog.pl (hunk 3 above)
Comment (npostavs) [reply .github/actions_scripts/getChangelog.pl:15]: Oh yep, done.

**Review by ann0see — APPROVED** [review]
Looks good.

File: .github/actions_scripts/getChangelog.pl (hunk 3 above)
Comment (ann0see) [reply .github/actions_scripts/getChangelog.pl:15]: Thanks!

**Pull request merged**